### PR TITLE
Implement try_chunks

### LINKED ISSUE #1947: try_chunks for futures 0.3
There is a [chunks](https://docs.rs/futures-preview/0.3.0-alpha.19/futures/stream/trait.StreamExt.html#method.chunks) method for `StreamExt`.

Could there be a `try_chunks` for `TryStreamExt`? The closure's domain would be `Vec<T>`, converting from `Result<T, E>`. 

One can produce the same result by performing the fold themselves, however it doesn't look and feel nice.

### PR DESCRIPTION
**Pull request by ivan770**
Closes #1947

### CODE REVIEW

**Review by taiki-e — COMMENTED** [review]
Thanks for the PR! I'm ok with adding such a feature, but I'm not sure if the behavior when an error occurs is good with this.

I feel it might be preferable to have our own error type that contains the error returned by the stream and the elements of the buffer at the time the error occurred.
What do you think?

**Comment by ivan770** [comment]
> 
> 
> Thanks for the PR! I'm ok with adding such a feature, but I'm not sure if the behavior when an error occurs is good with this.
> 
> I feel it might be preferable to have our own error type that contains the error returned by the stream and the elements of the buffer at the time the error occurred.
> What do you think?

Yeah, I thought about adding a tuple of all elements before error and the error itself, but my approach would lead to conversion of `TryStream` to `Stream`. I think your approach with custom error type seems more suitable, as `TryStream` is preserved. I'll modify PR a bit later.

**Comment by ivan770** [comment]
Sorry for lot's of intermediate commits and CI activation requests. I guess simple ID verification could be better for GitHub Actions 😅

File: futures-util/src/stream/try_stream/mod.rs [hunk 1]
```diff
@@ -506,6 +506,7 @@ pub trait TryStreamExt: TryStream {
     where
         Self: Sized,
     {
+        #[allow(clippy::type_complexity)]
```
Comment (ivan770) [inline futures-util/src/stream/try_stream/mod.rs:509]: I think it's ok to allow type complexity here, since it's still easy to read

**Review by taiki-e — APPROVED** [review]
Thanks!

**Comment by taiki-e** [comment]
Published in 0.3.16.

**Pull request merged**